xdsl/dialects/gpu.py

```diff
@@ -236,6 +249,62 @@ def get() -> LaneIdOp:
         return LaneIdOp.build(result_types=[IndexType()])
 
 
+@irdl_op_definition
+class LaunchOp(Operation):
+    name = "gpu.launch"
+    asyncDependencies: Annotated[VarOperand, AsyncTokenType]
+    gridSizeX: Annotated[Operand, IndexType]
+    gridSizeY: Annotated[Operand, IndexType]
+    gridSizeZ: Annotated[Operand, IndexType]
+    blockSizeX: Annotated[Operand, IndexType]
+    blockSizeY: Annotated[Operand, IndexType]
+    blockSizeZ: Annotated[Operand, IndexType]
+    dynamicSharedMemorySize: Annotated[OptOperand, i32]
+    asyncToken: Annotated[OptOpResult, AsyncTokenType]
+    body: Region
+    irdl_options = [AttrSizedOperandSegments()]
+
+    @staticmethod
+    def get(
+        body: Region,
+        gridSize: list[SSAValue | Operation],
+        blockSize: list[SSAValue | Operation],
+        async_launch: bool = False,
+        asyncDependencies: list[SSAValue | Operation] | None = None,
+        dynamicSharedMemorySize: SSAValue | Operation | None = None
+    ) -> LaunchOp:
+        if len(gridSize) != 3:
+            raise ValueError(
+                f"LaunchOp must have 3 gridSizes, got {len(gridSize)}")
+        if len(blockSize) != 3:
+            raise ValueError(
+                f"LaunchOp must have 3 blockSizes, got {len(blockSize)}")
+        operands = [[] if asyncDependencies is None else
+                    [SSAValue.get(a) for a in asyncDependencies]]
+
+        operands += [gs for gs in gridSize]
+        operands += [bs for bs in blockSize]
+        operands += [[] if dynamicSharedMemorySize is None else
+                     [SSAValue.get(dynamicSharedMemorySize)]]
+        return LaunchOp.build(
+            operands=operands,
+            result_types=[[AsyncTokenType()] if async_launch else []],
+            regions=[body])
+
+    def verify_(self) -> None:
+        if len(self.body.blocks) == 0 or all(
+            [len(b.ops) == 0 for b in self.body.blocks]):
+            raise VerifyException("gpu.launch requires a non-empty body.")
+        body_args = self.body.blocks[0].args
+        args_type = [a.typ for a in body_args]
+        if args_type != [IndexType()] * 12:
+            raise VerifyException(
+                f"Expected [12 x {str(IndexType())}], got {[str(t) for t in args_type]}."
```

Comment:
```suggestion
                f"Expected [12 x {str(IndexType())}], got {[str(t) for t in args_type]}. "
```

Gives a nicer error message.

Reply:
Applied, good catch, thanks!